plan.py

```diff
@@ -0,0 +1,53 @@
+from __future__ import annotations
+import multiprocessing as mp
+from dataclasses import dataclass, field, asdict, is_dataclass
+from typing import Any, Callable, Dict, List, Literal, Optional, Protocol, Set, Tuple, TYPE_CHECKING
+
+if TYPE_CHECKING:
+    from promise import ScopedConstraint
+
+Status = Literal["pending", "running", "success", "failed", "terminated"]
+
+@dataclass(slots=True)
+class TaskInstance:
+    id: str #UUID
+    spec_name: str
+    inputs: Dict[str, Any]
+    status: Status = "pending"
+    parent_id: Optional[str] = None
+    labels: Set[str] = field(default_factory=set)
+    started_at: Optional[float] = None
+    ended_at: Optional[float] = None
+    outputs: Any = None
+
+Edge = Tuple[str, str]  # (from_id, to_id)
+
+@dataclass(slots=True)
+class Plan:
+    nodes: Dict[str, TaskInstance] = field(default_factory=dict) #maps node_id to TaskIntance's
```

Comment:
```suggestion
    task_instances: Dict[str, TaskInstance] = field(default_factory=dict) #maps node_id to TaskIntance's
```

for readability since you have a class called `NewNode` below